### ISSUE TITLE
IndexError when computing CSA

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#1037]
data: folder_shared/sct_issues/20161122_csa

~~~
sct_process_segmentation -i t2_seg.nii.gz -p csa -vert 3:4

--
Spinal Cord Toolbox (master/73e29b174eb1ced5f3fd69ff34bcd5ea96675091)
Running /Users/julien/code/spinalcordtoolbox/scripts/sct_process_segmentation.py -i t2_seg.nii.gz -p csa -vert 3:4

Check parameters:
.. segmentation file:             t2_seg.nii.gz

Create temporary folder...
mkdir tmp.161122121826_753783/

Copying input data to tmp folder and convert to nii...
sct_convert -i /Users/julien/Dropbox/documents/processing/20161122_jisun/t2/t2_seg.nii.gz -o tmp.161122121826_753783/segmentation.nii.gz

Change orientation to RPI...
sct_image -i segmentation.nii.gz -setorient RPI -o segmentation_RPI.nii.gz

Open segmentation volume...

Get data dimensions...
  72 x 256 x 256

Smooth centerline/segmentation...
.. Get center of mass of the centerline/segmentation...
.. Smoothing algo = nurbs

Fitting centerline using B-spline approximation...
Test: # of control points = 5
Error on approximation = 11.18 mm
Test: # of control points = 6
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 7
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 8
Error on approximation = 0.85 mm
Test: # of control points = 9
Error on approximation = 0.33 mm
Test: # of control points = 10
Error on approximation = 0.21 mm
Test: # of control points = 11
Error on approximation = 0.12 mm
Test: # of control points = 12
Error on approximation = 0.1 mm
Test: # of control points = 13
Error on approximation = 0.07 mm
Test: # of control points = 14
Error on approximation = 0.07 mm
Test: # of control points = 15
Error on approximation = 0.05 mm
Test: # of control points = 16
Error on approximation = 0.05 mm
Test: # of control points = 17
Error on approximation = 0.05 mm
Number of control points of the optimal NURBS = 17

Compute CSA...

Smooth CSA across slices...
.. No smoothing!

Create volume of CSA values...

Create volume of angle values...
sct_image -i csa_volume_RPI.nii.gz -setorient AIL -o csa_volume_in_initial_orientation.nii.gz
sct_image -i angle_volume_RPI.nii.gz -setorient AIL -o angle_volume_in_initial_orientation.nii.gz

Generate output files...
  File created: /Users/julien/Dropbox/documents/processing/20161122_jisun/t2/csa_image.nii.gz
  File created: /Users/julien/Dropbox/documents/processing/20161122_jisun/t2/angle_image.nii.gz


Display CSA per slice:
z = 0, CSA = 45.7121967992 mm^2, Angle = 12.7885700378 deg
z = 1, CSA = 45.692658944 mm^2, Angle = 12.8960137286 deg
z = 2, CSA = 45.660511495 mm^2, Angle = 13.0709111159 deg
z = 3, CSA = 45.6249807609 mm^2, Angle = 13.2615780493 deg
z = 4, CSA = 45.5859870336 mm^2, Angle = 13.4677767477 deg
z = 5, CSA = 44.5941664554 mm^2, Angle = 13.6916702823 deg
z = 6, CSA = 43.602962164 mm^2, Angle = 13.9182645036 deg
z = 7, CSA = 44.5092601612 mm^2, Angle = 14.1325011352 deg
z = 8, CSA = 44.4692981636 mm^2, Angle = 14.3353843051 deg
z = 9, CSA = 45.3760440666 mm^2, Angle = 14.5286606665 deg
z = 10, CSA = 44.3961197165 mm^2, Angle = 14.6997972244 deg
z = 11, CSA = 47.1980141423 mm^2, Angle = 14.8465953486 deg
z = 12, CSA = 47.170984345 mm^2, Angle = 14.9698784103 deg
z = 13, CSA = 47.1479645782 mm^2, Angle = 15.0740947569 deg
z = 14, CSA = 46.1852641795 mm^2, Angle = 15.16471767 deg
z = 15, CSA = 45.2208448141 mm^2, Angle = 15.2665392194 deg
z = 16, CSA = 45.187250825 mm^2, Angle = 15.4217161688 deg
z = 17, CSA = 44.2018346742 mm^2, Angle = 15.6270036634 deg
z = 18, CSA = 44.1478684413 mm^2, Angle = 15.8751702999 deg
z = 19, CSA = 47.8370405614 mm^2, Angle = 16.1590711637 deg
z = 20, CSA = 46.8241745884 mm^2, Angle = 16.4718627709 deg
z = 21, CSA = 44.8726592666 mm^2, Angle = 16.8071547181 deg
z = 22, CSA = 44.7885555758 mm^2, Angle = 17.1591041237 deg
z = 23, CSA = 42.8374542954 mm^2, Angle = 17.5224634139 deg
z = 24, CSA = 45.5371792714 mm^2, Angle = 17.8925960163 deg
z = 25, CSA = 48.222614928 mm^2, Angle = 18.2654637307 deg
z = 26, CSA = 46.2675228306 mm^2, Angle = 18.6376041779 deg
z = 27, CSA = 44.3195579249 mm^2, Angle = 19.006101091 deg
z = 28, CSA = 48.8247639234 mm^2, Angle = 19.381253287 deg
z = 29, CSA = 48.7138207911 mm^2, Angle = 19.7480046411 deg
z = 30, CSA = 45.8564546057 mm^2, Angle = 20.0923396671 deg
z = 31, CSA = 47.5882596416 mm^2, Angle = 20.4261747794 deg
z = 32, CSA = 49.3104548231 mm^2, Angle = 20.7601667353 deg
z = 33, CSA = 50.1156837782 mm^2, Angle = 21.0825849849 deg
z = 34, CSA = 50.0133066241 mm^2, Angle = 21.3841314192 deg
z = 35, CSA = 50.816883317 mm^2, Angle = 21.6859815093 deg
z = 36, CSA = 53.4308870242 mm^2, Angle = 21.9756685655 deg
z = 37, CSA = 53.3266784202 mm^2, Angle = 22.2509515684 deg
z = 38, CSA = 53.2272645821 mm^2, Angle = 22.5105884665 deg
z = 39, CSA = 56.7359362391 mm^2, Angle = 22.752498645 deg
z = 40, CSA = 52.1473489741 mm^2, Angle = 22.9757071242 deg
z = 41, CSA = 53.8643022767 mm^2, Angle = 23.1783555063 deg
z = 42, CSA = 57.3770736193 mm^2, Angle = 23.3597811712 deg
z = 43, CSA = 53.726482236 mm^2, Angle = 23.518401306 deg
z = 44, CSA = 54.5655534922 mm^2, Angle = 23.6539025637 deg
z = 45, CSA = 58.9878771066 mm^2, Angle = 23.7650224713 deg
z = 46, CSA = 53.5885174992 mm^2, Angle = 23.8542215475 deg
z = 47, CSA = 58.0268625225 mm^2, Angle = 23.915225167 deg
z = 48, CSA = 58.9033185829 mm^2, Angle = 23.9508668713 deg
z = 49, CSA = 56.2202405275 mm^2, Angle = 23.9638353337 deg
z = 50, CSA = 58.8918413239 mm^2, Angle = 23.9759872808 deg
z = 51, CSA = 60.6642134663 mm^2, Angle = 24.0019400178 deg
z = 52, CSA = 56.1855814277 mm^2, Angle = 24.0431815692 deg
z = 53, CSA = 58.8364538112 mm^2, Angle = 24.0968681731 deg
z = 54, CSA = 60.5874352623 mm^2, Angle = 24.1642803027 deg
z = 55, CSA = 56.9859610578 mm^2, Angle = 24.2481399148 deg
z = 56, CSA = 59.612709153 mm^2, Angle = 24.3427842724 deg
z = 57, CSA = 54.2278106771 mm^2, Angle = 24.4509381873 deg
z = 58, CSA = 61.2782681855 mm^2, Angle = 24.5767407514 deg
z = 59, CSA = 60.3249280676 mm^2, Angle = 24.711752418 deg
z = 60, CSA = 62.0249827519 mm^2, Angle = 24.860114636 deg
z = 61, CSA = 62.8256992054 mm^2, Angle = 25.0273620582 deg
z = 62, CSA = 63.6161332277 mm^2, Angle = 25.2086470614 deg
z = 63, CSA = 61.7529101404 mm^2, Angle = 25.3971193912 deg
z = 64, CSA = 65.1721625045 mm^2, Angle = 25.5987681092 deg
z = 65, CSA = 64.1789459248 mm^2, Angle = 25.8072659772 deg
z = 66, CSA = 64.0870317276 mm^2, Angle = 25.9764367377 deg
z = 67, CSA = 64.8969393386 mm^2, Angle = 26.0992515282 deg
z = 68, CSA = 63.1016137044 mm^2, Angle = 26.175747479 deg
z = 69, CSA = 67.4661762083 mm^2, Angle = 26.2059487776 deg
z = 70, CSA = 64.8465825289 mm^2, Angle = 26.1898597871 deg
z = 71, CSA = 69.2651483409 mm^2, Angle = 26.1274623135 deg
z = 72, CSA = 67.5716664735 mm^2, Angle = 26.0233384643 deg
z = 73, CSA = 66.7779912228 mm^2, Angle = 25.8753568551 deg
z = 74, CSA = 68.6500704868 mm^2, Angle = 25.6768443907 deg
z = 75, CSA = 64.3809344188 mm^2, Angle = 25.4318193937 deg
z = 76, CSA = 66.3039936504 mm^2, Angle = 25.1402568732 deg
z = 77, CSA = 67.3662527296 mm^2, Angle = 24.8144565656 deg
z = 78, CSA = 66.6767881841 mm^2, Angle = 24.4447942298 deg
z = 79, CSA = 70.4691513585 mm^2, Angle = 24.0172551225 deg
z = 80, CSA = 71.6120032418 mm^2, Angle = 23.5611984796 deg
z = 81, CSA = 71.8728552618 mm^2, Angle = 23.0779311249 deg
z = 82, CSA = 71.2317558991 mm^2, Angle = 22.5854431489 deg
z = 83, CSA = 71.4877532782 mm^2, Angle = 22.0851564417 deg
z = 84, CSA = 73.5578099683 mm^2, Angle = 21.5782725821 deg
z = 85, CSA = 74.7264580978 mm^2, Angle = 21.0656124919 deg
z = 86, CSA = 72.2398423687 mm^2, Angle = 20.548621264 deg
z = 87, CSA = 74.309204518 mm^2, Angle = 20.046434238 deg
z = 88, CSA = 72.70410885 mm^2, Angle = 19.5426928727 deg
z = 89, CSA = 76.6207869046 mm^2, Angle = 19.0388823566 deg
z = 90, CSA = 78.7096709577 mm^2, Angle = 18.5187014226 deg
z = 91, CSA = 78.944091411 mm^2, Angle = 18.0022995909 deg
z = 92, CSA = 81.0245838685 mm^2, Angle = 17.5091978544 deg
z = 93, CSA = 84.0395809162 mm^2, Angle = 17.023711827 deg
z = 94, CSA = 84.2502511192 mm^2, Angle = 16.5481664303 deg
z = 95, CSA = 81.6348363387 mm^2, Angle = 16.0850651443 deg
z = 96, CSA = 85.5704165171 mm^2, Angle = 15.6553629087 deg
z = 97, CSA = 83.8466611864 mm^2, Angle = 15.2673451251 deg
z = 98, CSA = 85.8717881108 mm^2, Angle = 14.9183835227 deg
z = 99, CSA = 84.1000509597 mm^2, Angle = 14.6195606376 deg
z = 100, CSA = 83.2501774053 mm^2, Angle = 14.3662838152 deg
z = 101, CSA = 83.328559384 mm^2, Angle = 14.1541322775 deg
z = 102, CSA = 86.2332718122 mm^2, Angle = 13.9844820903 deg
z = 103, CSA = 87.233055866 mm^2, Angle = 13.8461486322 deg
z = 104, CSA = 85.3783765739 mm^2, Angle = 13.7321200596 deg
z = 105, CSA = 82.5628452276 mm^2, Angle = 13.6454474495 deg
z = 106, CSA = 79.7360349796 mm^2, Angle = 13.5855589237 deg
z = 107, CSA = 78.7993154439 mm^2, Angle = 13.5478340608 deg
z = 108, CSA = 79.7544290317 mm^2, Angle = 13.5307560415 deg
z = 109, CSA = 79.7527012949 mm^2, Angle = 13.5359128919 deg
z = 110, CSA = 82.5919133744 mm^2, Angle = 13.5621038943 deg
z = 111, CSA = 87.3207853564 mm^2, Angle = 13.610399111 deg
z = 112, CSA = 90.1397038353 mm^2, Angle = 13.6849920899 deg
z = 113, CSA = 90.1010071915 mm^2, Angle = 13.7856444527 deg
z = 114, CSA = 91.0076130484 mm^2, Angle = 13.8925111017 deg
z = 115, CSA = 90.9717385104 mm^2, Angle = 13.9835340887 deg
z = 116, CSA = 92.8358779002 mm^2, Angle = 14.0603916659 deg
z = 117, CSA = 93.7570049025 mm^2, Angle = 14.1241081329 deg
z = 118, CSA = 93.7379318587 mm^2, Angle = 14.1703548125 deg
z = 119, CSA = 92.7798657309 mm^2, Angle = 14.1977637336 deg
z = 120, CSA = 90.8824854842 mm^2, Angle = 14.2075122395 deg
z = 121, CSA = 90.885862536 mm^2, Angle = 14.1991006238 deg
z = 122, CSA = 89.0031316134 mm^2, Angle = 14.1717857401 deg
z = 123, CSA = 87.1270233355 mm^2, Angle = 14.1259331868 deg
z = 124, CSA = 87.1503345812 mm^2, Angle = 14.0648903513 deg
z = 125, CSA = 86.2319930919 mm^2, Angle = 13.9878932471 deg
z = 126, CSA = 87.2155363641 mm^2, Angle = 13.8927576029 deg
z = 127, CSA = 86.3080586334 mm^2, Angle = 13.7835391184 deg
z = 128, CSA = 93.944154634 mm^2, Angle = 13.6621911298 deg
z = 129, CSA = 91.1459324214 mm^2, Angle = 13.5359440055 deg
z = 130, CSA = 92.1438269428 mm^2, Angle = 13.4101439992 deg
z = 131, CSA = 93.136187274 mm^2, Angle = 13.3001874622 deg
z = 132, CSA = 93.1700766362 mm^2, Angle = 13.2117058115 deg
z = 133, CSA = 93.1957191305 mm^2, Angle = 13.1443671401 deg
z = 134, CSA = 89.4104593256 mm^2, Angle = 13.0930882964 deg
z = 135, CSA = 90.3755644064 mm^2, Angle = 13.055057389 deg
z = 136, CSA = 89.4328145843 mm^2, Angle = 13.0313508483 deg
z = 137, CSA = 87.5335561651 mm^2, Angle = 13.0212574218 deg
z = 138, CSA = 86.581282124 mm^2, Angle = 13.0236103604 deg
z = 139, CSA = 86.5758848004 mm^2, Angle = 13.0390431483 deg
z = 140, CSA = 86.5661171244 mm^2, Angle = 13.0669267244 deg
z = 141, CSA = 86.5514577043 mm^2, Angle = 13.1086652855 deg
z = 142, CSA = 87.4815486232 mm^2, Angle = 13.1676515473 deg
z = 143, CSA = 86.5034387389 mm^2, Angle = 13.2444807465 deg
z = 144, CSA = 82.6694014046 mm^2, Angle = 13.3374355191 deg
z = 145, CSA = 83.5799696893 mm^2, Angle = 13.4515640167 deg
z = 146, CSA = 86.3773210248 mm^2, Angle = 13.5948427878 deg
z = 147, CSA = 85.3684960531 mm^2, Angle = 13.7592276269 deg
z = 148, CSA = 85.3118578772 mm^2, Angle = 13.9136171132 deg
z = 149, CSA = 86.2065952701 mm^2, Angle = 14.0554772312 deg
z = 150, CSA = 84.2633080628 mm^2, Angle = 14.1868951105 deg
z = 151, CSA = 88.0063834945 mm^2, Angle = 14.2998162269 deg
z = 152, CSA = 88.9159567717 mm^2, Angle = 14.392344662 deg
z = 153, CSA = 88.8877140394 mm^2, Angle = 14.4630945635 deg
z = 154, CSA = 86.03267695 mm^2, Angle = 14.5099901399 deg
z = 155, CSA = 84.1322792777 mm^2, Angle = 14.5351475614 deg
z = 156, CSA = 83.1856201686 mm^2, Angle = 14.5387403233 deg
z = 157, CSA = 88.8649962327 mm^2, Angle = 14.519759298 deg
z = 158, CSA = 88.8812909431 mm^2, Angle = 14.4791376148 deg
z = 159, CSA = 87.9587840573 mm^2, Angle = 14.4208915165 deg
z = 160, CSA = 89.8808176243 mm^2, Angle = 14.3451936481 deg
z = 161, CSA = 89.9185502719 mm^2, Angle = 14.2508353405 deg
z = 162, CSA = 88.0674976604 mm^2, Angle = 14.1428767866 deg
z = 163, CSA = 87.1650060591 mm^2, Angle = 14.0263381593 deg
z = 164, CSA = 88.1539296647 mm^2, Angle = 13.9179632649 deg
z = 165, CSA = 87.2443685166 mm^2, Angle = 13.8159704922 deg
z = 166, CSA = 87.2819459502 mm^2, Angle = 13.7152599087 deg
z = 167, CSA = 87.317869089 mm^2, Angle = 13.618300116 deg
z = 168, CSA = 87.3513198238 mm^2, Angle = 13.5274005996 deg
z = 169, CSA = 87.3838610637 mm^2, Angle = 13.4383936366 deg
z = 170, CSA = 88.3669341011 mm^2, Angle = 13.3478382053 deg
z = 171, CSA = 92.2014306006 mm^2, Angle = 13.2590765091 deg
z = 172, CSA = 93.183864708 mm^2, Angle = 13.175539557 deg
z = 173, CSA = 94.165756909 mm^2, Angle = 13.0946000878 deg
z = 174, CSA = 93.244845515 mm^2, Angle = 13.0144024975 deg
z = 175, CSA = 94.2247943227 mm^2, Angle = 12.9392643901 deg
z = 176, CSA = 96.1537821166 mm^2, Angle = 12.8730570675 deg
z = 177, CSA = 93.3194315212 mm^2, Angle = 12.8146084213 deg
z = 178, CSA = 92.38566464 mm^2, Angle = 12.7641358086 deg
z = 179, CSA = 92.3991119025 mm^2, Angle = 12.7272693484 deg
z = 180, CSA = 90.5022138401 mm^2, Angle = 12.7041546799 deg
z = 181, CSA = 88.5996089147 mm^2, Angle = 12.696393049 deg
z = 182, CSA = 84.7879553113 mm^2, Angle = 12.6991454167 deg
z = 183, CSA = 83.8342264516 mm^2, Angle = 12.7023452881 deg
z = 184, CSA = 81.9294881034 mm^2, Angle = 12.7005301987 deg
z = 185, CSA = 81.9324632606 mm^2, Angle = 12.6912948652 deg
z = 186, CSA = 80.9845587491 mm^2, Angle = 12.6762099967 deg
z = 187, CSA = 80.9902028761 mm^2, Angle = 12.6584443153 deg
z = 188, CSA = 81.9478887608 mm^2, Angle = 12.6433054184 deg
z = 189, CSA = 81.949153599 mm^2, Angle = 12.6393624974 deg
z = 190, CSA = 81.9442857364 mm^2, Angle = 12.6545306179 deg
z = 191, CSA = 81.9295983952 mm^2, Angle = 12.7001879543 deg
z = 192, CSA = 80.9493678406 mm^2, Angle = 12.7864306886 deg
z = 193, CSA = 81.8591953988 mm^2, Angle = 12.9168399918 deg
z = 194, CSA = 83.7006144545 mm^2, Angle = 13.1013096287 deg
z = 195, CSA = 86.463899896 mm^2, Angle = 13.3552928269 deg
z = 196, CSA = 89.1987753973 mm^2, Angle = 13.6641067346 deg
z = 197, CSA = 90.9504577188 mm^2, Angle = 14.0372555825 deg
z = 198, CSA = 93.6041845019 mm^2, Angle = 14.4905927329 deg
z = 199, CSA = 96.2218690145 mm^2, Angle = 14.9853224287 deg
z = 200, CSA = 102.741445007 mm^2, Angle = 15.1589632343 deg
z = 201, CSA = 104.824509891 mm^2, Angle = 14.7536510767 deg
z = 202, CSA = 110.8811844 mm^2, Angle = 13.9640592733 deg
z = 203, CSA = 116.898774867 mm^2, Angle = 13.2932269357 deg
z = 204, CSA = 122.126953275 mm^2, Angle = 14.2008147153 deg
z = 205, CSA = 120.061511452 mm^2, Angle = 18.9653503465 deg
z = 206, CSA = 109.873872581 mm^2, Angle = 29.2873736573 deg
z = 207, CSA = 41.8383660993 mm^2, Angle = 45.3852886609 deg
z = 208, CSA = 10.930862273 mm^2, Angle = 65.508195883 deg
z = 209, CSA = 0.24976871143 mm^2, Angle = 82.6528059634 deg
Save results in: /Users/julien/Dropbox/documents/processing/20161122_jisun/t2/csa_per_slice.txt

Save results in: /Users/julien/Dropbox/documents/processing/20161122_jisun/t2/csa_per_slice.pickle

Selected vertebral levels... 3:4
  OK: ./label/template/PAM50_levels.nii.gz
WARNING: File ./label/template/PAM50_levels.nii.gz already exists. Deleting it.
	Find slices corresponding to vertebral levels based on the centerline...
Traceback (most recent call last):
  File "/Users/julien/code/spinalcordtoolbox/scripts/sct_process_segmentation.py", line 1067, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/spinalcordtoolbox/scripts/sct_process_segmentation.py", line 231, in main
    compute_csa(fname_segmentation, output_folder, overwrite, verbose, remove_temp_files, step, smoothing_param, figure_fit, slices, vert_lev, fname_vertebral_labeling, algo_fitting = param.algo_fitting, type_window= param.type_window, window_length=param.window_length, angle_correction=angle_correction)
  File "/Users/julien/code/spinalcordtoolbox/scripts/sct_process_segmentation.py", line 646, in compute_csa
    slices, vert_levels_list, warning = get_slices_matching_with_vertebral_levels_based_centerline(vert_levels, im_vertebral_labeling.data, z_centerline_fit_vox)
  File "/Users/julien/code/spinalcordtoolbox/scripts/sct_process_segmentation.py", line 941, in get_slices_matching_with_vertebral_levels_based_centerline
    vertebral_labeling_slice_iz = vertebral_labeling_data[:, :, z_centerline[i_z]]
IndexError: index 256 is out of bounds for axis 2 with size 256
~~~

**Comment by jcohenadad** [comment]
another strange issue: why does it say at the end of the terminal:
~~~
WARNING: File ./label/template/PAM50_levels.nii.gz already exists. Deleting it.
~~~

This file is NOT supposed to be deleted!

**Comment by slevyrosetti** [comment]
The problem came from deprecated code that was not deleted when fixing issues related to transformation from physical to voxel coordinates. It seems that several issues are caused by this kind of manipulations. With George, we suggest for the future that both the data in physical coordinates and voxel coordinates (and maybe also in fitted (real) coordinates) are recorded under different variable names instead of going from one type of coordinates to another, and vice versa.

**Comment by slevyrosetti** [comment]
Fixed in branch `sl_process_seg_issue1037`.

**Comment by jcohenadad** [comment]
great, could you please do a pull request?

**Comment by jcohenadad** [comment]
how about my comment related to the file that is deleted?

**Comment by slevyrosetti** [comment]
The file was not actually deleted, it was copied in a temporary folder and its orientation was then changed into RPI. I have now changed this to use the functions of the class `msct_image` to avoid loading the data and so on, it is now faster and does not display this message any longer.

**Comment by jcohenadad** [comment]
fixed in PR #1040
